src/goto-symex/symex_main.cpp

```diff
@@ -67,17 +85,50 @@ void symex_transition(
     // 1. the transition from state.source.pc to "to" is not a backwards goto
     // or
     // 2. we are arriving from an outer loop
+
+    // TODO: This should all be replaced by natural loop analysis.
     framet &frame = state.call_stack().top();
     const goto_programt::instructiont &instruction=*to;
     for(const auto &i_e : instruction.incoming_edges)
-      if(i_e->is_goto() && i_e->is_backwards_goto() &&
-         (!is_backwards_goto ||
-          state.source.pc->location_number>i_e->location_number))
+    {
+      if(
+        i_e->is_goto() && i_e->is_backwards_goto() &&
+        (!is_backwards_goto ||
+         state.source.pc->location_number > i_e->location_number))
       {
         const auto loop_id =
           goto_programt::loop_id(state.source.function_id, *i_e);
-        frame.loop_iterations[loop_id].count = 0;
+        auto &current_loop_info = frame.loop_iterations[loop_id];
+        current_loop_info.count = 0;
+
+        // We've found a loop, put it on the stack and say it's our current
+        // active loop.
+        if(frame.loops_info)
+          frame.active_loops.emplace_back(frame.loops_info->loop_map[to]);
+      }
+    }
+
+    // Otherwise if we find we're transitioning out of a loop, make sure
+    // to remove any loops we're not currently iterating over.
+
+    // Match the do-while pattern.
+    if(
+      state.source.pc->is_backwards_goto() &&
+      state.source.pc->location_number < to->location_number)
+    {
```

Comment:
I don't think you need the pattern recognition any more? Can just always call the pop method?

Reply:
Nope, still need it. Lexical loop analysis alone isn't isn't accurate enough by itself, even it does help.